main.c

```diff
@@ -621,6 +622,117 @@ static void tcmur_stop_device(void *arg)
 	tcmu_dev_dbg(dev, "cmdproc cleanup done\n");
 }
 
+int tcmur_get_time(struct tcmu_device *dev, struct timespec *time)
+{
+	struct tcmur_device *rdev = tcmu_dev_get_private(dev);
+	int ret;
+
+	ret = clock_gettime(CLOCK_MONOTONIC_COARSE, time);
+	if (!ret) {
+		tcmu_dev_dbg(dev, "Current time %lu secs.\n", time->tv_sec);
+		return 0;
+	}
+
+	tcmu_dev_err(dev, "Could not get time. Error %d. Command timeout feature disabled.\n",
+		     ret);
+	rdev->cmd_time_out = 0;
+	return ret;
+}
+
+static bool get_next_cmd_timeout(struct tcmu_device *dev,
+				 struct timespec *curr_time,
+				 struct timespec *tmo)
+{
+	struct tcmur_device *rdev = tcmu_dev_get_private(dev);
+	int left, cmd_tmo = rdev->cmd_time_out;
+	struct tcmur_cmd *tcmur_cmd;
+	bool has_timeout = false;
+
+	if (!cmd_tmo)
+		return false;
+
+	memset(tmo, 0, sizeof(*tmo));
+
+	pthread_spin_lock(&rdev->lock);
+	list_for_each(&rdev->cmds_list, tcmur_cmd, cmds_list_entry) {
+		if (tcmur_cmd->timed_out)
+			continue;
+
+		has_timeout = true;
+		/* mnc - Rollover? */
+		left = curr_time->tv_sec - tcmur_cmd->start_time.tv_sec;
+		if (cmd_tmo > left) {
+			tmo->tv_sec = cmd_tmo -left;
+		} else {
+			/*
+			 * We do not do a clock call for every command, so
+			 * cmds can time out while we were processing new
+			 * cmds. Force a recheck.
+			 */
+			tmo->tv_sec = 0;
+		}
+
+		tcmu_dev_dbg(dev, "Next cmd id %hu timeout in %lu secs. Current time %lu. Start time %lu\n",
+			     tcmur_cmd->lib_cmd->cmd_id, tmo->tv_sec,
+			     curr_time->tv_sec, tcmur_cmd->start_time.tv_sec);
+		break;
+	}
+	pthread_spin_unlock(&rdev->lock);
+
+	return has_timeout;
+}
+
+static void check_for_timed_out_cmds(struct tcmu_device *dev)
+{
+	struct tcmur_device *rdev = tcmu_dev_get_private(dev);
+	int cmd_tmo = rdev->cmd_time_out;
+	struct tcmur_cmd *tcmur_cmd;
+	struct timespec curr_time;
+
+	if (!cmd_tmo)
+		return;
+
+	memset(&curr_time, 0, sizeof(curr_time));
+
+	if (tcmur_get_time(dev, &curr_time))
+		return;
+
+	pthread_spin_lock(&rdev->lock);
+	list_for_each(&rdev->cmds_list, tcmur_cmd, cmds_list_entry) {
+		if (tcmur_cmd->timed_out)
+			continue;
+
+		/* mnc - Rollover? */
+		if (tcmur_cmd->start_time.tv_sec + cmd_tmo > curr_time.tv_sec)
+			continue;
+
+		tcmu_dev_info(dev, "Command %hu timed out.\n",
+			      tcmur_cmd->lib_cmd->cmd_id);
```

Comment:
Here we may also need to print out the whole cdb info, which will be very useful to get what kind of the SCSI commands get stuck, here the cmd_id is only  for the ID in the ringbuffer.

This sometimes will also give us a clear clue to debug it further in the backend storage service.

Reply:
Ok. I will add that.

Reply:
I added this for the case SCSI CDB DEBUG is set. For the INFO case, I also added some more info: SCSI cdb[0] value, the LBA and length.

Reply:
Yeah, it makes sense and cool.
I will test it right now.
Thanks Mike.